py/desitarget/mock/selection.py

```diff
@@ -213,25 +273,60 @@ def ndens_select(data, source_name, **kwargs):
     else:
         in_z = z>0.0
 
-    try:
-        bounds = kwargs['min_ra'], kwargs['max_ra'], kwargs['min_dec'], kwargs['max_dec']
-        mock_dens = estimate_density(ra[in_z], dec[in_z], bounds=bounds)
-    except KeyError:
-        mock_dens = estimate_density(ra[in_z], dec[in_z])
-
-    frac_keep = min(kwargs['number_density']/mock_dens , 1.0)
-    if mock_dens < kwargs['number_density']:
-        print("WARNING: mock cannot achieve the goal density for source {} Goal {}. Mock {}".format(source_name, kwargs['number_density'], mock_dens))
-
-
     n = len(ra)
-    keepornot = np.random.uniform(0.,1.,n)
-    limit = np.zeros(n) + frac_keep
-    kept = keepornot < limit
-    select_sample = (in_z) & (kept)
 
+    limit = np.zeros(n) 
     target_class = np.zeros(n,dtype=np.int64) - 1
-    target_class[select_sample] = desi_mask.mask(source_name)
+    keepornot = np.random.uniform(0.,1.,n)
+
+    if density_fluctuations:
+        bricks = desispec.brick.brickname(ra, dec)
+        unique_bricks = list(set(bricks))
+    
+        for brickname in unique_bricks:
+            in_brick = (brickname == bricks)
+            n_in_brick = np.count_nonzero(in_brick)
+   #         print('brickname {} len {}'.format(brickname, n_in_brick))
+        
+            #locate the brick info we need
+            id_binfo  = np.where(brick_info['BRICKNAME'] == brickname)
+            id_binfo = id_binfo[0]
+
+            if len(id_binfo) != 1:
+                num_density = 0.0
+                brick_area = 0.25
+                warnings.warn("Tile is on the border. NumDensity= 0.0", RuntimeWarning)
+            else:
+                brick_area = brick_info['BRICKAREA'][id_binfo]
+                num_density = brick_info['FLUC_EBV'][sourcename][id_binfo]  * brick_info['NTARGET_'+sourcename] 
+                
+                
+            mock_dens = n_in_brick/brick_area
+                
+            frac_keep = num_density/mock_dens
+  #          print('num density desired {}, num density in mock {}, frac_keep {}'.format(num_density, mock_dens, frac_keep))
+            
+            kept = keepornot < (limit + frac_keep)
```

Comment:
Since `limit` is just an array of zeros, I don't think it is needed here, and you could just use `kept = keepornot < frac_keep`.  It should be ok to do the boolean comparison between an array (`keepornot`) and a scalar (`frac_keep`).